include/drivers/gpio.h

```diff
@@ -233,6 +543,75 @@ static inline int gpio_pin_read(struct device *port, u32_t pin,
 	return gpio_read(port, GPIO_ACCESS_BY_PIN, pin, value);
 }
 
+/**
+ * @brief Set pin logical level.
+ *
+ * Set logical value of a pin taking into account GPIO_ACTIVE_LOW flag.
+ * If pin is configured as Active Low, writing value 1 will set it to a low
+ * physical level. Otherwise, writing value 1 will set it to a high physical
+ * level.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param pin Pin number where the data is written.
+ * @param value Value set on the pin.
+ */
+__syscall void gpio_pin_set(struct device *port, u32_t pin, bool value);
+
+static inline void z_impl_gpio_pin_set(struct device *port, u32_t pin,
+				       bool value)
+{
+	const struct gpio_driver_api *api =
+		(const struct gpio_driver_api *)port->driver_api;
+
+	return api->pin_set(port, pin, value);
+}
+
+/**
+ * @brief Get pin logical level.
+ *
+ * Get logical value of a pin taking into account GPIO_ACTIVE_LOW flag.
+ * If pin is configured as Active Low, a low physical level will be interpreted
+ * as logical value 1. Otherwise, a low physical level will be interpreted as
+ * logical value 0.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param pin Pin number where data is read.
+ *
+ * @return pin logical value.
+ */
+__syscall bool gpio_pin_get(struct device *port, u32_t pin);
+
+static inline bool z_impl_gpio_pin_get(struct device *port, u32_t pin)
+{
+	const struct gpio_driver_api *api =
+		(const struct gpio_driver_api *)port->driver_api;
+
+	return api->pin_get(port, pin);
+}
+
+/**
+ * @brief Configure pin interrupt.
+ *
+ * @param port Pointer to device structure for the driver instance.
+ * @param pin Pin number to configure.
+ * @param flags Interrupt configuration flags as defined by GPIO_INT_*.
```

Comment:
My first concern here is that this suggests that only flags that begin with `GPIO_INT_*` can appear in the `flags` parameter.  This implies that those flags should *not* appear in the `flags` parameter to `gpio_pin_configure`, but that is not documented.

The problem is that when dealing with a pin that is used both in non-interrupt (output chip-select, input polling) and an interrupt (device needs service) I can imagine situations where other flags might be desired.  A specific example is that I might need a pull-up on an active-low interrupt, but it's not part of the gpio configuration flags for use as an output.

So: which flags are permitted in which API?

Also, in the telecon we discussed whether configuring a pin for interrupt use should block use of the gpio API on it (e.g. you can't write to a pin that's used for interrupts).  I don't see any documentation on how interoperability between the two modes, nor implementation that prevents some operations.  (At a minimum enabling callbacks on a pin that's not configured for interrupts should be rejected, or something.)

Finally, this configures (essentially enables) interrupts separately from the callbacks, but doesn't take into account discussion from #17102 which proposes combining the configure, enable, and callback configuration into a single function that places a (or several) pins into interrupt mode.  I don't think this function should be added without considering how it will interact with that sort of refactoring (which, incidentally, would resolve some of the questions above).

Reply:
I've pushed a new gpio.h version where description of flags parameter explicitly mentions which flags are allowed. Also added a comment clarifying that `gpio_pin_configure` should always be called first. Any other function in the GPIO API, including `gpio_pin_interrupt_configure` should be called only after the main pin configuration has been done. `gpio_pin_interrupt_configure` doesn't introduce a new working mode, only adds interrupt capability on top of an already existing pin configuration. If pin is configured as an output only the `gpio_pin_interrupt_configure` should report it via assert. The description of what checks are performed via assert is not yet added. I will do it next.

Reply:
> Any other function in the GPIO API, including gpio_pin_interrupt_configure should be called only after the main pin configuration has been done.

I assume gpio_pin_configure ("main pin configuration") muxes the pin to GPIO. Requiring this before you can use gpio_pin_interrupt_configure prevents to set interrupt callbacks on pins that are muxed to alternate functions as SPI, I2C, ... In my view a showstopper. This is also an indication that the concept of interrupt on pin level changes and pin characteristic configuration should be handled independently (see e.g. https://elinux.org/images/b/b6/Pin_Control_Subsystem_Overview.pdf page 17/18). 

> If pin is configured as an output only the gpio_pin_interrupt_configure should report it via assert.

This should not be a general rule. There are SoCs that allow to generate an interrupt even when configured as an output.

Reply:
The general idea is that the future pinctrl API will allow application to connect peripheral to a pin as well as configure pin properties: drive strength, pull up/down, open source drive mode, etc. The application will not need GPIO module to do that. I was planning to use the same approach for the interrupt configuration but actually interrupts are different. It makes sense to enforce application to configure interrupts via GPIO module. In this case the statement
> Any other function in the GPIO API, including gpio_pin_interrupt_configure should be called only after the main pin configuration has been done.

is not valid. I'll modify it to clarify that `gpio_pin_interrupt_configure()` can be called on the pin controlled by a peripheral module other than GPIO. Only documentation will change. The implementation stays as it is now.

Modifying properties of a pin controlled by a peripheral module other than GPIO will still be possible only after we introduce pinctrl driver.

> > If pin is configured as an output only the gpio_pin_interrupt_configure should report it via assert.
> 
> This should not be a general rule. There are SoCs that allow to generate an interrupt even when configured as an output.

Could you give an example of such a SoC? Please notice that my comment states "output only". Having a pin configured as input-output (or input) is fine. Typically, if the input path is disconnected in GPIO hardware module there is no way to generate interrupts as interrupt subsystem is blind, unaware of pin changes.

Reply:
> Could you give an example of such a SoC? Please notice that my comment states "output only". Having a pin configured as input-output (or input) is fine.

Sorry, I misinterpreted your statement "output **only**" and took it for "output" without the input disconnected.